### PR TITLE
[DD-963] Solr schema parser

### PR DESCRIPTION
**Pull request by boferri**
- finc Solr schema as inbuilt d:swarm schema

### CODE REVIEW

File: converter/src/main/java/org/dswarm/converter/schema/SolrSchemaParser.java [hunk 1]
```diff
@@ -0,0 +1,372 @@
...
+	}
+
+	public Optional<Schema> parse(final String solrSchemaFilePath, final String schemaUUID, final String schemaName)
+			throws DMPPersistenceException {
+
+		final Optional<Document> optionalDocument = readXML(solrSchemaFilePath);
+
+		if (!optionalDocument.isPresent()) {
+
+			LOG.error("parsed Solr schema (from '" + solrSchemaFilePath + "') is not present");
```
Comment (knutwalker) [inline converter/src/main/java/org/dswarm/converter/schema/SolrSchemaParser.java:111]: `LOG.error("parsed Solr schema (from '{}') is not present", solrSchemaFilePath);`

File: converter/src/main/java/org/dswarm/converter/schema/SolrSchemaParser.java [hunk 2]
```diff
@@ -0,0 +1,372 @@
...
+			return Optional.empty();
+		}
+
+		final Document document = optionalDocument.get();
+
+		final Optional<NodeList> optionalFields = getFields(document, solrSchemaFilePath);
+
+		if (!optionalFields.isPresent()) {
+
+			LOG.error("couldn't find fields in the Solr schema (from '" + solrSchemaFilePath + "')");
```
Comment (knutwalker) [inline converter/src/main/java/org/dswarm/converter/schema/SolrSchemaParser.java:122]: `LOG.error("couldn't find fields in the Solr schema (from '{}')", solrSchemaFilePath);`

File: converter/src/main/java/org/dswarm/converter/schema/SolrSchemaParser.java [hunk 3]
```diff
@@ -0,0 +1,372 @@
...
+
+		final Document document;
+
+		try {
+			document = documentBuilder.parse(inputStream);
+
+			return Optional.ofNullable(document);
+		} catch (final SAXException e) {
+
+			LOG.error("couldn't parse the Solr schema file at '" + solrSchemaFilePath + "'", e);
```
Comment (knutwalker) [inline converter/src/main/java/org/dswarm/converter/schema/SolrSchemaParser.java:197]: `LOG.error("couldn't parse the Solr schema file at '{}'", solrSchemaFilePath, e);`

File: converter/src/main/java/org/dswarm/converter/schema/SolrSchemaParser.java [hunk 4]
```diff
@@ -0,0 +1,372 @@
...
+
+			return Optional.empty();
+		} catch (final IOException e) {
+
+			LOG.error("couldn't read the Solr schema fila at '" + solrSchemaFilePath + "'", e);
```
Comment (knutwalker) [inline converter/src/main/java/org/dswarm/converter/schema/SolrSchemaParser.java:202]: `LOG.error("couldn't read the Solr schema fila at '{}'", solrSchemaFilePath, e);`

File: converter/src/main/java/org/dswarm/converter/schema/SolrSchemaParser.java [hunk 5]
```diff
@@ -0,0 +1,372 @@
...
+		}
+
+		final NodeList result;
+
+		try {
+
+			result = (NodeList) expr.evaluate(document, XPathConstants.NODESET);
+		} catch (final XPathExpressionException e) {
+
+			LOG.error("could not execute xpath query against document (from '" + solrSchemaFilePath + "')", e);
```
Comment (knutwalker) [inline converter/src/main/java/org/dswarm/converter/schema/SolrSchemaParser.java:231]: `LOG.error("could not execute xpath query against document (from '{}')", solrSchemaFilePath, e);`

File: converter/src/main/java/org/dswarm/converter/schema/SolrSchemaParser.java [hunk 6]
```diff
@@ -0,0 +1,372 @@
...
+		final List<Attribute> attributes = new ArrayList<>();
+
+		// determine and mint attributes
+		for (int i = 0; i < fields.getLength(); i++) {
+
+			final Node field = fields.item(i);
+
+			if (!field.hasAttributes()) {
+
+				LOG.debug("field ('" + ToStringBuilder.reflectionToString(field) + "') has no attributes, cannot parse an attribute from it");
```
Comment (knutwalker) [inline converter/src/main/java/org/dswarm/converter/schema/SolrSchemaParser.java:250]:
```
if (LOG.isDebugEnabled()) {
    LOG.debug("field ('{}') has no attributes, cannot parse an attribute from it", ToStringBuilder.reflectionToString(field));
}
```

File: converter/src/main/java/org/dswarm/converter/schema/SolrSchemaParser.java [hunk 7]
```diff
@@ -0,0 +1,372 @@
...
+				continue;
+			}
+
+			final NamedNodeMap fieldAttributes = field.getAttributes();
+
+			final Optional<Node> optionalNameNode = Optional.ofNullable(fieldAttributes.getNamedItem(NAME_IDENTIFIER));
+
+			if (!optionalNameNode.isPresent()) {
+
+				LOG.debug("field ('" + ToStringBuilder.reflectionToString(field) + "') has no name XML attribute, cannot parse an attribute from it");
```
Comment (knutwalker) [inline converter/src/main/java/org/dswarm/converter/schema/SolrSchemaParser.java:261]:
```
if (LOG.isDebugEnabled()) {
    LOG.debug("field ('{}') has no name XML attribute, cannot parse an attribute from it", ToStringBuilder.reflectionToString(field));
}
```

File: converter/src/main/java/org/dswarm/converter/schema/SolrSchemaParser.java [hunk 8]
```diff
@@ -0,0 +1,372 @@
...
+				continue;
+			}
+
+			final Node nameNode = optionalNameNode.get();
+
+			final Optional<String> optionalName = Optional.ofNullable(nameNode.getNodeValue());
+
+			if (!optionalName.isPresent()) {
+
+				LOG.debug("field ('" + ToStringBuilder.reflectionToString(field) + "') has no name, cannot parse an attribute from it");
```
Comment (knutwalker) [inline converter/src/main/java/org/dswarm/converter/schema/SolrSchemaParser.java:272]:
```
if (LOG.isDebugEnabled()) {
    LOG.debug("field ('{}') has no name, cannot parse an attribute from it", ToStringBuilder.reflectionToString(field));
}
```

File: converter/src/test/resources/finc-solr-schema.xml [hunk 9]
```diff
@@ -0,0 +1,363 @@
+<?xml version="1.0" ?>
```
Comment (knutwalker) [inline converter/src/test/resources/finc-solr-schema.xml:1]: Is it OK for us to include this file here? I think, UBL should at least consent to this.

Comment (boferri) [inline converter/src/test/resources/finc-solr-schema.xml:1]: dunno, I'll ask them ;)

**Comment by boferri** [comment]
@knutwalker https://github.com/zazi/dswarm/commit/88c94377e8997fef0af27168f9e8dba411b8463e is part of PR #38

**Pull request merged**